File: 백준/Gold/1068. 트리/트리.cc
#include <bits/stdc++.h>
using namespace std;

int n, node, root, r;
vector<int> v[54];

int dfs(int here){
    int ret = 0;
    int child = 0;
    for(int there : v[here]){
        if(there == r) continue;
        ret += dfs(there);
        child++;
    }
    if(child == 0) return 1;
    return ret;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin >> n;
    for(int i = 0; i < n; i++){
        cin >> node;
        if(node == -1) root = i;
        else v[node].push_back(i);
    }
    cin >> r;
    if(r == root){
        cout << 0 << '\n';
        return 0;
    }
    else{
        cout << dfs(root) << '\n';
    }
    return 0;
}

File: 백준/Gold/1068. 트리/트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

class Main {
    static ArrayList<Integer>[] nodeList;
    static int n, root, deleteN;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        nodeList = new ArrayList[54];
        for(int i = 0; i < 54; i++){
            nodeList[i] = new ArrayList<>();
        }
        for(int i = 0; i < n; i++){
            int parent = Integer.parseInt(st.nextToken());
            if(parent == -1)
                root = i;
            else {
                nodeList[parent].add(i);
            }
        }
        deleteN = Integer.parseInt(br.readLine());
        if(deleteN == root)
            System.out.println(0);
        else {
            System.out.println(dfs(root));
        }
    }

    private static int dfs(int start) {
        int ret = 0;
        int child = 0;
        for(int adj : nodeList[start]){
            if(adj == deleteN) continue;
            ret += dfs(adj);
            child++;
        }
        if(child == 0)
            return 1;
        return ret;
    }
}